static_frame/core/index_hierarchy.py

```diff
@@ -456,211 +719,273 @@ def _from_type_blocks(cls: tp.Type[IH],
+            ]
+            self._indexers = indices._indexers
+            self._name = name if name is not NAME_DEFAULT else indices._name
+            self._blocks = indices._blocks
+            self._values = indices._values
+            self._map = indices._map
+            return
+
+        if not (indexers.__class__ is np.ndarray and not indexers.flags.writeable):
+            raise ErrorInitIndex('indexers must be a read-only numpy array.')
+
+        if not all(isinstance(index, Index) for index in indices):
+            raise ErrorInitIndex("indices must all Index's!")
+
+        if len(indices) <= 1:
+            raise ErrorInitIndex('Index Hierarchies must have at least two levels!')
+
+        self._indices = [
+            mutable_immutable_index_filter(self.STATIC, index) # type: ignore
+            for index in indices
+        ]
+        self._indexers = indexers
+        self._name = None if name is NAME_DEFAULT else name_filter(name)
 
+        if blocks is not None:
+            if own_blocks:
+                self._blocks = blocks
+            else:
+                self._blocks = blocks.copy()
         else:
-            raise NotImplementedError(f'no handling for creation from {levels}')
+            self._blocks = self._create_blocks_from_self()
 
-        if self.STATIC and index_level.STATIC:
-            self._levels = index_level
-        else: # must deepcopy IndexLevels if not IndexHierarchy not static
-            self._levels = index_level.to_index_level(
-                    cls=self._LEVEL_CONSTRUCTOR
-                    )
+        self._values = self._blocks.values
+        self._map = HierarchicalLocMap(indices=self._indices, indexers=self._indexers)
 
-        if self._levels.depth <= 1:
-            raise ErrorInitIndex(f'invalid depth ({self._levels.depth}) for IndexLevels composed in IndexHierarchy')
+    def _update_array_cache(self: IH) -> None:
+        # This MUST be set before entering this context
+        assert self._pending_extensions is not None
 
-        self._recache = self._blocks is None
-        self._name = None if name is NAME_DEFAULT else name_filter(name)
+        new_indexers = [np.empty(self.__len__(), DTYPE_INT_DEFAULT) for _ in range(self.depth)]
 
-    #---------------------------------------------------------------------------
-    def __deepcopy__(self: IH, memo: tp.Dict[int, tp.Any]) -> IH:
+        current_size = len(self._blocks)
+
+        for depth, indexer in enumerate(self._indexers):
+            new_indexers[depth][:current_size] = indexer
+
+        # TODO-Ariza: Should we clear self._indexers?
+        # TODO-Ariza: What if an error is raised anywhere onward in this method?
+
+        offset = current_size
+
+        # For all these extensions, we have already update self._indices - we now need to map indexers
+        for pending in self._pending_extensions: # pylint: disable = E1133
+            if isinstance(pending, PendingRow):
```

Comment:
No chance of subclass, right?
```suggestion
            if pending.__class__ is PendingRow:
```